src/lib/ares_process.c

```diff
@@ -816,6 +856,68 @@ static struct server_state *ares__random_server(ares_channel_t *channel)
   return NULL;
 }
 
+/* Pick a server from the list with failover behavior.
+ *
+ * We default to using the first server in the sorted list of servers. That is
+ * the server with the lowest number of consecutive failures and then the
+ * highest priority server (by idx) if there is a draw.
+ *
+ * However, if a server temporarily goes down and hits some failures, then that
+ * server will never be retried until all other servers hit the same number of
+ * failures. This may prevent the server from being retried for a long time.
+ *
+ * To resolve this, with some probability we select a failed server to retry
+ * instead.
+ */
+static struct server_state *ares__failover_server(ares_channel_t *channel)
+{
+  struct server_state *first_server = ares__slist_first_val(channel->servers);
+  struct server_state *last_server  = ares__slist_last_val(channel->servers);
+  unsigned short       r;
+
+  /* Defensive code against no servers being available on the channel. */
+  if (first_server == NULL) {
+    return NULL;
+  }
+
+  /* If no servers have failures, then prefer the first server in the list. */
+  if (last_server != NULL && last_server->consec_failures == 0) {
+    return first_server;
+  }
+
+  /* If we are not configured with a server retry chance then return the first
+   * server.
+   */
+  if (channel->server_retry_chance == 0) {
+    return first_server;
+  }
+
+  /* Generate a random value to decide whether to retry a failed server. The
+   * probability to use is 1/channel->server_retry_chance, rounded up to a
+   * precision of 1/2^B where B is the number of bits in the random value.
+   * We use an unsigned short for the random value for increased precision.
+   */
+  ares__rand_bytes(channel->rand_state, (unsigned char *)&r, sizeof(r));
+  if (r % channel->server_retry_chance == 0) {
+    /* Select a suitable failed server to retry. */
+    struct timeval      now = ares__tvnow();
+    ares__slist_node_t *node;
+    for (node = ares__slist_node_first(channel->servers); node != NULL;
+         node = ares__slist_node_next(node)) {
+      struct server_state *node_val = ares__slist_node_val(node);
+      if (node_val != NULL && node_val->consec_failures > 0 &&
+          ares__timedout(&now, &node_val->next_retry_time)) {
+        timeadd(&now, channel->server_retry_delay);
+        node_val->next_retry_time = now;
```

Comment:
This should be set in `server_increment_failures()` not here. Each time the server is tried and failed, it should push out the retry delay, regardless of if it was selected due to random selection or not (e.g. the highest priority server failed and its next in the priority order). 

This also mitigates the need to modify the test cases as it appears the current code path is allowing an *immediate* retry of the server without honoring the timeout.

It does make sense to add test cases for coverage for this scenario however, and that can be done using an ultra-low retry timeout and high probability, but that should be a separate test case and not modification of the existing one.

Reply:
Thanks, I definitely agree with that change and it makes the FV behaviour a lot more reliable.

I have added another set of test cases (and left the existing tests as-is), which cover 4 scenarios for server retries.
- No retries if all servers are healthy.
- Simple retry after the retry delay has passed.
- Failed servers are retried in sorted order (i.e. number of consec failures and then index).
- Failed servers which have not yet met the retry delay are skipped and not retried.

I set the retry to delay to 100ms in these tests, and call `usleep` to increment time. There is a trade off between reliability and speed of the tests, but I felt that 100ms was appropriate.